File: src/main/java/net/gtaun/shoebill/samp/SampCallbackManagerImpl.java
/**
 * Copyright (C) 2011-2014 MK124
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.samp;

import net.gtaun.shoebill.util.TryUtils;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * @author MK124 & 123marvin123
 */
public class SampCallbackManagerImpl implements SampCallbackManager {
    private Queue<SampCallbackHandler> callbackHandlers;
    private SampCallbackHandler callbackHandler = new SampCallbackHandler() {
        @Override
        public void onProcessTick() {
            for (SampCallbackHandler handler : callbackHandlers) {
                try {
                    handler.onProcessTick();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        @Override
        public void onShoebillUnload() {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(handler::onShoebillUnload));
        }

        @Override
        public void onShoebillLoad() {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(handler::onShoebillLoad));
        }

        @Override
        public void onAmxLoad(int handle) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onAmxLoad(handle)));
        }

        @Override
        public void onAmxUnload(int handle) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onAmxUnload(handle)));
        }

        @Override
        public boolean onGameModeInit() {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(handler::onGameModeInit));
            return true;
        }

        @Override
        public boolean onGameModeExit() {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(handler::onGameModeExit));
            return true;
        }

        @Override
        public boolean onPlayerConnect(int playerid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerConnect(playerid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerDisconnect(int playerid, int reason) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerDisconnect(playerid, reason)));
            return ret[0];
        }

        @Override
        public boolean onPlayerSpawn(int playerid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerSpawn(playerid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerDeath(int playerid, int killerid, int reason) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerDeath(playerid, killerid, reason)));
            return true;
        }

        @Override
        public boolean onVehicleSpawn(int vehicleid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleSpawn(vehicleid)));
            return true;
        }

        @Override
        public boolean onVehicleDeath(int vehicleid, int killerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleDeath(vehicleid, killerid)));
            return true;
        }

        @Override
        public int onPlayerText(int playerid, String text) {
            int[] ret = {1};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerText(playerid, text)));
            return ret[0];
        }

        @Override
        public int onPlayerCommandText(int playerid, String cmdtext) {
            int[] ret = {0};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] |= handler.onPlayerCommandText(playerid, cmdtext)));
            return ret[0];
        }

        @Override
        public boolean onPlayerRequestClass(int playerid, int classid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerRequestClass(playerid, classid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerEnterVehicle(int playerid, int vehicleid, int ispassenger) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerEnterVehicle(playerid, vehicleid, ispassenger)));
            return true;
        }

        @Override
        public boolean onPlayerExitVehicle(int playerid, int vehicleid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerExitVehicle(playerid, vehicleid)));
            return true;
        }

        @Override
        public boolean onPlayerStateChange(int playerid, int newstate, int oldstate) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerStateChange(playerid, newstate, oldstate)));
            return true;
        }

        @Override
        public boolean onPlayerEnterCheckpoint(int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerEnterCheckpoint(playerid)));
            return true;
        }

        @Override
        public boolean onPlayerLeaveCheckpoint(int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerLeaveCheckpoint(playerid)));
            return true;
        }

        @Override
        public boolean onPlayerEnterRaceCheckpoint(int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerEnterRaceCheckpoint(playerid)));
            return true;
        }

        @Override
        public boolean onPlayerLeaveRaceCheckpoint(int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerLeaveRaceCheckpoint(playerid)));
            return true;
        }

        @Override
        public int onRconCommand(String cmd) {
            int[] ret = {0};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] |= handler.onRconCommand(cmd)));
            return ret[0];
        }

        @Override
        public boolean onPlayerRequestSpawn(int playerid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerRequestSpawn(playerid)));
            return ret[0];
        }

        @Override
        public boolean onObjectMoved(int objectid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onObjectMoved(objectid)));
            return true;
        }

        @Override
        public boolean onPlayerObjectMoved(int playerid, int objectid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerObjectMoved(playerid, objectid)));
            return true;
        }

        @Override
        public boolean onPlayerPickUpPickup(int playerid, int pickupid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerPickUpPickup(playerid, pickupid)));
            return true;
        }

        @Override
        public boolean onVehicleMod(int playerid, int vehicleid, int componentid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onVehicleMod(playerid, vehicleid, componentid)));
            return ret[0];
        }

        @Override
        public boolean onEnterExitModShop(int playerid, int enterexit, int interiorid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onEnterExitModShop(playerid, enterexit, interiorid)));
            return true;
        }

        @Override
        public boolean onVehiclePaintjob(int playerid, int vehicleid, int paintjobid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehiclePaintjob(playerid, vehicleid, paintjobid)));
            return true;
        }

        @Override
        public boolean onVehicleRespray(int playerid, int vehicleid, int color1, int color2) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onVehicleRespray(playerid, vehicleid, color1, color2)));
            return ret[0];
        }

        @Override
        public boolean onVehicleDamageStatusUpdate(int vehicleid, int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleDamageStatusUpdate(vehicleid, playerid)));
            return true;
        }

        @Override
        public boolean onUnoccupiedVehicleUpdate(int vehicleid, int playerid, int passengerSeat, float newX, float newY, float newZ, float vel_x, float vel_y, float vel_z) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onUnoccupiedVehicleUpdate(vehicleid, playerid, passengerSeat, newX, newY, newZ, vel_x, vel_y, vel_z)));
            return ret[0];
        }

        @Override
        public boolean onPlayerSelectedMenuRow(int playerid, int row) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerSelectedMenuRow(playerid, row)));
            return true;
        }

        @Override
        public boolean onPlayerExitedMenu(int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerExitedMenu(playerid)));
            return true;
        }

        @Override
        public boolean onPlayerInteriorChange(int playerid, int newinteriorid, int oldinteriorid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerInteriorChange(playerid, newinteriorid, oldinteriorid)));
            return true;
        }

        @Override
        public boolean onPlayerKeyStateChange(int playerid, int newkeys, int oldkeys) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerKeyStateChange(playerid, newkeys, oldkeys)));
            return ret[0];
        }

        @Override
        public int onRconLoginAttempt(String ip, String password, int success) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onRconLoginAttempt(ip, password, success)));
            return 1;
        }

        @Override
        public boolean onPlayerUpdate(int playerid) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerUpdate(playerid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerStreamIn(int playerid, int forplayerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerStreamIn(playerid, forplayerid)));
            return true;
        }

        @Override
        public boolean onPlayerStreamOut(int playerid, int forplayerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerStreamOut(playerid, forplayerid)));
            return true;
        }

        @Override
        public boolean onVehicleStreamIn(int vehicleid, int forplayerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleStreamIn(vehicleid, forplayerid)));
            return true;
        }

        @Override
        public boolean onVehicleStreamOut(int vehicleid, int forplayerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleStreamOut(vehicleid, forplayerid)));
            return true;
        }

        @Override
        public int onDialogResponse(int playerid, int dialogid, int response, int listitem, String inputtext) {
            int[] ret = {0};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] |= handler.onDialogResponse(playerid, dialogid, response, listitem, inputtext)));
            return ret[0];
        }

        @Override
        public boolean onPlayerTakeDamage(int playerId, int issuerId, float amount, int weaponId, int bodypart) {
            boolean[] ret = {false};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerTakeDamage(playerId, issuerId, amount, weaponId, bodypart)));
            return ret[0];
        }

        @Override
        public boolean onPlayerGiveDamage(int playerId, int damagedId, float amount, int weaponId, int bodypart) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerGiveDamage(playerId, damagedId, amount, weaponId, bodypart)));
            return true;
        }

        @Override
        public boolean onPlayerClickMap(int playerId, float x, float y, float z) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerClickMap(playerId, x, y, z)));
            return true;
        }

        @Override
        public boolean onPlayerClickTextDraw(int playerid, int clickedid) {
            boolean[] ret = {false};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] |= handler.onPlayerClickTextDraw(playerid, clickedid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerClickPlayerTextDraw(int playerid, int playertextid) {
            boolean[] ret = {false};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] |= handler.onPlayerClickPlayerTextDraw(playerid, playertextid)));
            return ret[0];
        }

        @Override
        public boolean onPlayerClickPlayer(int playerid, int clickedplayerid, int source) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerClickPlayer(playerid, clickedplayerid, source)));
            return true;
        }

        @Override
        public int onPlayerEditObject(int playerid, int playerobject, int objectid, int response, float fX, float fY, float fZ, float fRotX, float fRotY, float fRotZ) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerEditObject(playerid, playerobject, objectid, response, fX, fY, fZ, fRotX, fRotY, fRotZ)));
            return 1;
        }

        @Override
        public boolean onPlayerEditAttachedObject(int playerid, int response, int index, int modelid, int boneid, float fOffsetX, float fOffsetY, float fOffsetZ, float fRotX, float fRotY, float fRotZ, float fScaleX, float fScaleY, float fScaleZ) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerEditAttachedObject(playerid, response, index, modelid, boneid, fOffsetX, fOffsetY, fOffsetZ, fRotX, fRotY, fRotZ, fScaleX, fScaleY, fScaleZ)));
            return true;
        }

        @Override
        public boolean onPlayerSelectObject(int playerid, int type, int objectid, int modelid, float fX, float fY, float fZ) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerSelectObject(playerid, type, objectid, modelid, fX, fY, fZ)));
            return true;
        }

        @Override
        public boolean onPlayerWeaponShot(int playerid, int weaponid, int hittype, int hitid, float fX, float fY, float fZ) {
            boolean[] ret = {true};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onPlayerWeaponShot(playerid, weaponid, hittype, hitid, fX, fY, fZ)));
            return ret[0];
        }

        @Override
        public int onIncomingConnection(int playerid, String ipAddress, int port) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onIncomingConnection(playerid, ipAddress, port)));
            return 1;
        }

        @Override
        public boolean onTrailerUpdate(int playerid, int vehicleid) {
            boolean[] ret = {false};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> ret[0] &= handler.onTrailerUpdate(playerid, vehicleid)));
            return ret[0];
        }

        @Override
        public int[] onHookCall(String name, Object... objects) {
            Object[] event = {new int[]{0, 0}};
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> event[0] = handler.onHookCall(name, objects)));
            return (int[]) event[0];
        }

        @Override
        public boolean onActorStreamIn(int actor, int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onActorStreamIn(actor, playerid)));
            return true;
        }

        @Override
        public boolean onActorStreamOut(int actor, int playerid) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onActorStreamOut(actor, playerid)));
            return true;
        }

        @Override
        public int onPlayerGiveDamageActor(int playerid, int actor, int amount, int weapon, int bodypart) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onPlayerGiveDamageActor(playerid, actor, amount, weapon, bodypart)));
            return 1;
        }

        @Override
        public int onVehicleSirenStateChange(int playerid, int vehicleid, int newstate) {
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> handler.onVehicleSirenStateChange(playerid, vehicleid, newstate)));
            return 1;
        }

        @Override
        public int onRegisteredFunctionCall(int amx, String name, Object[] parameters) {
            int[] returnValue = new int[] { -1 };
            callbackHandlers.forEach(handler -> TryUtils.tryTo(() -> returnValue[0] = handler.onRegisteredFunctionCall(amx, name, parameters)));
            return returnValue[0];
        }
    };

    public SampCallbackManagerImpl() {
        callbackHandlers = new ConcurrentLinkedQueue<>();
    }

    @Override
    public void registerCallbackHandler(SampCallbackHandler handler) {
        callbackHandlers.add(handler);
    }

    @Override
    public void unregisterCallbackHandler(SampCallbackHandler handler) {
        callbackHandlers.remove(handler);
    }

    @Override
    public boolean hasCallbackHandler(SampCallbackHandler handler) {
        return callbackHandlers.contains(handler);
    }

    @Override
    public SampCallbackHandler getMasterCallbackHandler() {
        return callbackHandler;
    }
}

File: src/main/java/net/gtaun/shoebill/samp/SampCallbackManagerImpl.kt
/**
 * Copyright (C) 2011-2014 MK124

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.samp

import net.gtaun.shoebill.ShoebillImpl
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * @author MK124
 * @author Marvin Haschker
 */
class SampCallbackManagerImpl : SampCallbackManager {
    private val callbackHandlers: Queue<SampCallbackHandler> = ConcurrentLinkedQueue<SampCallbackHandler>()
    override val masterCallbackHandler: SampCallbackHandler = object : SampCallbackHandler {

        override fun onProcessTick() = callbackHandlers.forEachCatched { it.onProcessTick() }
        override fun onShoebillUnload() = callbackHandlers.forEachCatched { it.onShoebillUnload() }
        override fun onShoebillLoad() = callbackHandlers.forEachCatched { it.onShoebillLoad() }
        override fun onAmxLoad(handle: Int) = callbackHandlers.forEachCatched { it.onAmxLoad(handle) }
        override fun onAmxUnload(handle: Int) = callbackHandlers.forEachCatched { it.onAmxUnload(handle) }

        override fun onGameModeInit(): Boolean {
            callbackHandlers.forEachCatched { it.onGameModeInit() }
            return true
        }

        override fun onGameModeExit(): Boolean {
            callbackHandlers.forEachCatched { it.onGameModeExit() }
            return true
        }

        override fun onPlayerConnect(playerId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerConnect(playerId) }
            return ret
        }

        override fun onPlayerDisconnect(playerId: Int, reason: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerDisconnect(playerId, reason) }
            return ret
        }

        override fun onPlayerSpawn(playerId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerSpawn(playerId) }
            return ret
        }

        override fun onPlayerDeath(playerId: Int, killerId: Int, reason: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerDeath(playerId, killerId, reason) }
            return true
        }

        override fun onVehicleSpawn(vehicleId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleSpawn(vehicleId) }
            return true
        }

        override fun onVehicleDeath(vehicleId: Int, killerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleDeath(vehicleId, killerId) }
            return true
        }

        override fun onPlayerText(playerId: Int, text: String): Int {
            var ret = 1
            callbackHandlers.forEachCatched { ret = it.onPlayerText(playerId, text) }
            return ret
        }

        override fun onPlayerCommandText(playerId: Int, commandText: String): Int {
            var ret = 0
            callbackHandlers.forEachCatched { ret = it.onPlayerCommandText(playerId, commandText) }
            return ret
        }

        override fun onPlayerRequestClass(playerId: Int, classId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerRequestClass(playerId, classId) }
            return ret
        }

        override fun onPlayerEnterVehicle(playerId: Int, vehicleId: Int, isPassenger: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerEnterVehicle(playerId, vehicleId, isPassenger) }
            return true
        }

        override fun onPlayerExitVehicle(playerId: Int, vehicleId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerExitVehicle(playerId, vehicleId) }
            return true
        }

        override fun onPlayerStateChange(playerId: Int, state: Int, oldState: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerStateChange(playerId, state, oldState) }
            return true
        }

        override fun onPlayerEnterCheckpoint(playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerEnterCheckpoint(playerId) }
            return true
        }

        override fun onPlayerLeaveCheckpoint(playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerLeaveCheckpoint(playerId) }
            return true
        }

        override fun onPlayerEnterRaceCheckpoint(playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerEnterRaceCheckpoint(playerId) }
            return true
        }

        override fun onPlayerLeaveRaceCheckpoint(playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerLeaveRaceCheckpoint(playerId) }
            return true
        }

        override fun onRconCommand(cmd: String): Int {
            var ret = 0
            callbackHandlers.forEachCatched { ret = it.onRconCommand(cmd) }
            return ret
        }

        override fun onPlayerRequestSpawn(playerId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerRequestSpawn(playerId) }
            return ret
        }

        override fun onObjectMoved(objectId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onObjectMoved(objectId) }
            return true
        }

        override fun onPlayerObjectMoved(playerId: Int, objectId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerObjectMoved(playerId, objectId) }
            return true
        }

        override fun onPlayerPickUpPickup(playerId: Int, pickupId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerPickUpPickup(playerId, pickupId) }
            return true
        }

        override fun onVehicleMod(playerId: Int, vehicleId: Int, componentId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onVehicleMod(playerId, vehicleId, componentId) }
            return ret
        }

        override fun onEnterExitModShop(playerId: Int, enterexit: Int, interiorId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onEnterExitModShop(playerId, enterexit, interiorId) }
            return true
        }

        override fun onVehiclePaintjob(playerId: Int, vehicleId: Int, paintjobId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehiclePaintjob(playerId, vehicleId, paintjobId) }
            return true
        }

        override fun onVehicleRespray(playerId: Int, vehicleId: Int, color1: Int, color2: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onVehicleRespray(playerId, vehicleId, color1, color2) }
            return ret
        }

        override fun onVehicleDamageStatusUpdate(vehicleId: Int, playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleDamageStatusUpdate(vehicleId, playerId) }
            return true
        }

        override fun onUnoccupiedVehicleUpdate(vehicleId: Int, playerId: Int, passengerSeat: Int, newX: Float,
                                               newY: Float, newZ: Float, velX: Float, velY: Float,
                                               velZ: Float): Boolean {
            var ret = true
            callbackHandlers.forEachCatched {
                ret = it.onUnoccupiedVehicleUpdate(vehicleId, playerId,
                    passengerSeat, newX, newY, newZ, velX, velY, velZ) }
            return ret
        }

        override fun onPlayerSelectedMenuRow(playerId: Int, row: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerSelectedMenuRow(playerId, row) }
            return true
        }

        override fun onPlayerExitedMenu(playerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerExitedMenu(playerId) }
            return true
        }

        override fun onPlayerInteriorChange(playerId: Int, interiorId: Int, oldInteriorId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerInteriorChange(playerId, interiorId, oldInteriorId) }
            return true
        }

        override fun onPlayerKeyStateChange(playerId: Int, keys: Int, oldKeys: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerKeyStateChange(playerId, keys, oldKeys) }
            return ret
        }

        override fun onRconLoginAttempt(ip: String, password: String, isSuccess: Int): Int {
            callbackHandlers.forEachCatched { it.onRconLoginAttempt(ip, password, isSuccess) }
            return 1
        }

        override fun onPlayerUpdate(playerId: Int): Boolean {
            var ret = true
            callbackHandlers.forEachCatched { ret = it.onPlayerUpdate(playerId) }
            return ret
        }

        override fun onPlayerStreamIn(playerId: Int, forPlayerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerStreamIn(playerId, forPlayerId) }
            return true
        }

        override fun onPlayerStreamOut(playerId: Int, forPlayerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerStreamOut(playerId, forPlayerId) }
            return true
        }

        override fun onVehicleStreamIn(vehicleId: Int, forPlayerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleStreamIn(vehicleId, forPlayerId) }
            return true
        }

        override fun onVehicleStreamOut(vehicleId: Int, forPlayerId: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleStreamOut(vehicleId, forPlayerId) }
            return true
        }

        override fun onDialogResponse(playerId: Int, dialogId: Int, response: Int, listitem: Int, inputtext: String): Int {
            var ret = 0
            callbackHandlers.forEachCatched {
                ret = it.onDialogResponse(playerId, dialogId, response,
                    listitem, inputtext) }
            return ret
        }

        override fun onPlayerTakeDamage(playerId: Int, issuerId: Int, amount: Float, weaponId: Int, bodypart: Int): Boolean {
            var ret = false
            callbackHandlers.forEachCatched { ret = it.onPlayerTakeDamage(playerId, issuerId, amount, weaponId, bodypart) }
            return ret
        }

        override fun onPlayerGiveDamage(playerId: Int, damagedId: Int, amount: Float, weaponId: Int, bodypart: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerGiveDamage(playerId, damagedId, amount, weaponId, bodypart) }
            return true
        }

        override fun onPlayerClickMap(playerId: Int, x: Float, y: Float, z: Float): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerClickMap(playerId, x, y, z) }
            return true
        }

        override fun onPlayerClickTextDraw(playerId: Int, clickedId: Int): Boolean {
            var ret = false
            callbackHandlers.forEachCatched { ret = it.onPlayerClickTextDraw(playerId, clickedId) }
            return ret
        }

        override fun onPlayerClickPlayerTextDraw(playerId: Int, playerTextId: Int): Boolean {
            var ret = false
            callbackHandlers.forEachCatched { ret = it.onPlayerClickPlayerTextDraw(playerId, playerTextId) }
            return ret
        }

        override fun onPlayerClickPlayer(playerId: Int, clickedPlayerId: Int, source: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerClickPlayer(playerId, clickedPlayerId, source) }
            return true
        }

        override fun onPlayerEditObject(playerId: Int, playerObject: Int, objectId: Int, response: Int, fX: Float,
                                        fY: Float, fZ: Float, fRotX: Float, fRotY: Float, fRotZ: Float): Int {
            callbackHandlers.forEachCatched {
                it.onPlayerEditObject(playerId, playerObject, objectId, response, fX, fY,
                    fZ, fRotX, fRotY, fRotZ) }
            return 1
        }

        override fun onPlayerEditAttachedObject(playerId: Int, response: Int, index: Int, modelid: Int, boneid: Int,
                                                fOffsetX: Float, fOffsetY: Float, fOffsetZ: Float, fRotX: Float,
                                                fRotY: Float, fRotZ: Float, fScaleX: Float, fScaleY: Float,
                                                fScaleZ: Float): Boolean {
            callbackHandlers.forEachCatched {
                it.onPlayerEditAttachedObject(playerId, response, index, modelid, boneid,
                    fOffsetX, fOffsetY, fOffsetZ, fRotX, fRotY, fRotZ, fScaleX, fScaleY, fScaleZ) }
            return true
        }

        override fun onPlayerSelectObject(playerId: Int, type: Int, objectId: Int, modelId: Int, fX: Float, fY: Float,
                                          fZ: Float): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerSelectObject(playerId, type, objectId, modelId, fX, fY, fZ) }
            return true
        }

        override fun onPlayerWeaponShot(playerId: Int, weaponId: Int, hitType: Int, hitId: Int, fX: Float, fY: Float,
                                        fZ: Float): Boolean {
            var ret = true
            callbackHandlers.forEachCatched {
                ret = it.onPlayerWeaponShot(playerId, weaponId, hitType,
                    hitId, fX, fY, fZ) }
            return ret
        }

        override fun onIncomingConnection(playerId: Int, ipAddress: String, port: Int): Int {
            callbackHandlers.forEachCatched { it.onIncomingConnection(playerId, ipAddress, port) }
            return 1
        }

        override fun onTrailerUpdate(playerId: Int, vehicleId: Int): Boolean {
            var ret = false
            callbackHandlers.forEachCatched { ret = it.onTrailerUpdate(playerId, vehicleId) }
            return ret
        }


        override fun onRegisteredFunctionCall(amx: Int, name: String, parameters: Array<Any>): Int {
            var ret = -1
            callbackHandlers.forEachCatched { ret = it.onRegisteredFunctionCall(amx, name, parameters) }
            return ret
        }

        override fun onVehicleSirenStateChange(playerid: Int, vehicleid: Int, newstate: Int): Boolean {
            callbackHandlers.forEachCatched { it.onVehicleSirenStateChange(playerid, vehicleid, newstate) }
            return true
        }

        override fun onPlayerGiveDamageActor(playerid: Int, actor: Int, amount: Float, weapon: Int, bodypart: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerGiveDamageActor(playerid, actor, amount, weapon, bodypart) }
            return true
        }

        override fun onActorStreamOut(actor: Int, playerid: Int): Boolean {
            callbackHandlers.forEachCatched { it.onActorStreamOut(actor, playerid) }
            return true
        }

        override fun onActorStreamIn(actor: Int, playerid: Int): Boolean {
            callbackHandlers.forEachCatched { it.onActorStreamIn(actor, playerid) }
            return true
        }

        override fun onHookCall(name: String, vararg objects: Any): IntArray {
            var array = intArrayOf(0, 0)
            callbackHandlers.forEachCatched { array = it.onHookCall(name, *objects) }
            return array
        }

        override fun onPlayerFinishedDownloading(playerId: Int, virtualWorld: Int): Boolean {
            callbackHandlers.forEachCatched { it.onPlayerFinishedDownloading(playerId, virtualWorld) }
            return true
        }
    }

    override fun registerCallbackHandler(handler: SampCallbackHandler) {
        callbackHandlers.add(handler)
    }

    override fun unregisterCallbackHandler(handler: SampCallbackHandler) {
        callbackHandlers.remove(handler)
    }

    override fun hasCallbackHandler(handler: SampCallbackHandler): Boolean = callbackHandlers.contains(handler)
}

private fun <E> Queue<E>.forEachCatched(any: (E) -> Unit) {
    this.forEach {
        try {
            any.invoke(it)
        } catch (e: Throwable) {
            e.printStackTrace()
            ShoebillImpl.instance.reportError(e)
        }
    }
}
